apollo/apollo.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2023 CloudWeGo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apollo
+
+import (
+	"bytes"
+	"errors"
+	"text/template"
+
+	"github.com/apolloconfig/agollo/v4/component/log"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/shima-park/agollo"
+)
+
+// Client the wrapper of apollo client.
+type Client interface {
+	SetParser(ConfigParser)
+	ClientConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (ConfigParam, error)
+	ServerConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (ConfigParam, error)
+	RegisterConfigCallback(ConfigParam, func(string, ConfigParser))
+	DeregisterConfig(ConfigParam) error
+}
+
+type ConfigParam struct {
+	Key       string
+	NameSpace string
+	Cluster   string
+	Content   string
+	Type      ConfigType
+}
+
+type client struct {
+	acli agollo.Agollo
+	// support customise parser
+	parser            ConfigParser
+	clusterTemplate   *template.Template
+	namespaceTemplate *template.Template
+	serverKeyTemplate *template.Template
+	clientKeyTemplate *template.Template
+}
+
+const (
+	RetryConfigName          = "retry"
+	RpcTimeoutConfigName     = "rpc_timeout"
+	CircuitBreakerConfigName = "circuit_break"
+
+	LimiterConfigName = "limit"
+)
+
+type Options struct {
+	ConfigServerURL string
+	NamespaceID     string
+	AppID           string
+	Cluster         string
+	ServerKeyFormat string
+	ClientKeyFormat string
+	IsPrivate       bool
+	AccessKey       string
+	BackupFilePath  string
+	CustomLogger    log.LoggerInterface
+	ConfigParser    ConfigParser
+}
+
+func New(opts Options) (Client, error) {
+	if opts.ConfigServerURL == "" {
+		opts.ConfigServerURL = ApolloDefaultConfigServerURL
+	}
+	if opts.CustomLogger == nil {
+		opts.CustomLogger = NewCustomApolloLogger()
+	}
+	if opts.ConfigParser == nil {
+		opts.ConfigParser = defaultConfigParse()
+	}
+	if opts.AppID == "" {
+		opts.AppID = ApolloDefaultAppId
+	}
+	// TODO
+	if opts.NamespaceID == "" {
+		opts.NamespaceID = ApolloNameSpace
+	}
+	if opts.Cluster == "" {
+		opts.Cluster = ApolloDefaultCluster
+	}
+	if opts.ServerKeyFormat == "" {
+		opts.ServerKeyFormat = ApolloDefaultServerKey
+	}
+	if opts.ClientKeyFormat == "" {
+		opts.ClientKeyFormat = ApolloDefaultClientKey
+	}
+	agolloOption := []agollo.Option{
+		agollo.Cluster(opts.Cluster),
+		agollo.PreloadNamespaces([]string{RetryConfigName, RpcTimeoutConfigName, CircuitBreakerConfigName, LimiterConfigName}...),
```

Comment:
Since the names of the namespaces are already hardcoded, it's meaningless to allow user customization.

Reply:
The preload namespace is for init config before config listening, there is a example in line 219 of the code file, the option of AutoFetchOnCacheMiss maybe solved it , it uses the interface with cache, but just in case ,I configured the preloaded namespace